File: app/src/main/java/com/qimian233/ztool/hook/modules/packageinstaller/packageInstallerStyleHook.java
package com.qimian233.ztool.hook.modules.packageinstaller;

import com.qimian233.ztool.hook.base.BaseHookModule;
import io.github.libxposed.api.XposedInterface;
import io.github.libxposed.api.XposedModuleInterface;

import java.lang.reflect.Field;
import java.lang.reflect.Method;

import android.app.Activity;
import android.app.AlertDialog;
import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.widget.TextView;

/**
 * ZUI包安装器Hook模块
 * 功能：绕过ZUI系统的安装限制，修改包安装器界面样式
 * 目标：com.android.packageinstaller (ZUI系统包安装器)
 */
public class packageInstallerStyleHook extends BaseHookModule {

    public packageInstallerStyleHook() {}

    @Override
    public String getModuleName() {
        return "packageInstallerStyle_hook";
    }

    @Override
    public String[] getTargetPackages() {
        return new String[]{
                "com.android.packageinstaller"  // ZUI系统包安装器
        };
    }

    @Override
    public void handleLoadPackage(XposedModuleInterface.PackageLoadedParam param) throws Throwable {
        ClassLoader classLoader = param.getDefaultClassLoader();
        String packageName = param.getPackageName();
        if ("com.android.packageinstaller".equals(packageName)) {
            hookZuiPackageInstaller(classLoader);
            doNotShowWarnTextHook(classLoader);
        }
    }

    private void hookZuiPackageInstaller(ClassLoader classLoader) {
        try {
            // 1. Hook Utils类的isCTSandGTS方法，绕过安装限制
            hookInstallationRestrictions(classLoader);

            // 2. Hook Activity样式，修改界面显示
            hookActivityStyles(classLoader);

            log("ZUI Package Installer Hook 成功加载");
        } catch (Throwable t) {
            logError("ZUI Package Installer Hook 加载失败", t);
        }
    }

    private void hookInstallationRestrictions(ClassLoader classLoader) {
        try {
            Class<?> utilsClass = classLoader.loadClass(
                    "com.android.packageinstaller.extra.Utils");

            // Hook isCTSandGTS方法的重载版本
            Method isCTSandGTS1 = utilsClass.getDeclaredMethod("isCTSandGTS", String.class);
            this.xposed.hook(isCTSandGTS1).intercept(chain -> Boolean.TRUE);

            Method isCTSandGTS2 = utilsClass.getDeclaredMethod("isCTSandGTS", String.class, Intent.class);
            this.xposed.hook(isCTSandGTS2).intercept(chain -> Boolean.TRUE);

            log("成功Hook安装限制检查方法");
        } catch (Throwable t) {
            logError("Hook安装限制检查方法失败", t);
        }
    }

    private void hookActivityStyles(ClassLoader classLoader) {
        try {
            // 获取Theme_AlertDialogActivity的资源ID
            Class<?> styleClass = classLoader.loadClass(
                    "com.android.packageinstaller.R$style");
            Field themeField = styleClass.getDeclaredField("Theme_AlertDialogActivity");
            themeField.setAccessible(true);
            final int themeAlertDialogActivity = themeField.getInt(null);

            // Hook Activity的onCreate方法，修改主题和窗口属性
            Method onCreate = Activity.class.getDeclaredMethod("onCreate", Bundle.class);
            this.xposed.hook(onCreate).intercept(chain -> {
                Activity activity = (Activity) chain.getThisObject();

                // 检查是否为目标包安装器的Activity
                if (activity.getPackageName().equals("com.android.packageinstaller")) {
                    try {
                        // 设置对话框主题
                        activity.setTheme(themeAlertDialogActivity);

                        // 设置透明背景
                        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
                            activity.setTranslucent(true);
                        }

                        // 请求无标题栏
                        activity.requestWindowFeature(1); // 1对应Window.FEATURE_NO_TITLE

                        // 禁用窗口动画
                        activity.getWindow().setWindowAnimations(0);

                        log("成功修改包安装器Activity样式");
                    } catch (Throwable t) {
                        logError("修改Activity样式时出错", t);
                    }
                }

                return chain.proceed();
            });

            log("成功Hook Activity样式修改");
        } catch (Throwable t) {
            logError("Hook Activity样式修改失败", t);
        }
    }

    private void doNotShowWarnTextHook(ClassLoader classLoader) {
        try {
            Class<?> activityClass = classLoader.loadClass(
                    "com.android.packageinstaller.PackageInstallerActivity");
            Method startInstallConfirm = activityClass.getDeclaredMethod("startInstallConfirm");
            this.xposed.hook(startInstallConfirm).intercept(chain -> {
                Object result = chain.proceed();
                try {
                    Class<?> resourcesClass = classLoader.loadClass("com.android.packageinstaller.R$id");
                    Field warnTextViewIdField = resourcesClass.getDeclaredField("install_confirm_question_warning");
                    warnTextViewIdField.setAccessible(true);
                    int warnTextViewId = warnTextViewIdField.getInt(null);

                    Field mDialogField = chain.getThisObject().getClass().getDeclaredField("mDialog");
                    mDialogField.setAccessible(true);
                    AlertDialog dialog = (AlertDialog) mDialogField.get(chain.getThisObject());

                    TextView tv = dialog.findViewById(warnTextViewId);
                    tv.setVisibility(TextView.GONE);
                    log("Successfully set install warn visibility to GONE");
                } catch (Exception e) {
                    logError("Exception happened when trying to set warn text to GONE!", e);
                }
                return result;
            });
        } catch (Throwable t) {
            logError("Failed to hook doNotShowWarnTextHook", t);
        }
    }
}

File: app/src/main/java/com/qimian233/ztool/hook/modules/packageinstaller/PackageInstallerStyleHook.kt
package com.qimian233.ztool.hook.modules.packageinstaller

import android.annotation.SuppressLint
import android.app.Activity
import android.app.AlertDialog
import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.widget.TextView
import com.qimian233.ztool.data.keys.PreferenceKeys
import com.qimian233.ztool.data.keys.ScopeKeys
import com.qimian233.ztool.hook.base.AppHookModule
import io.github.libxposed.api.XposedModuleInterface.PackageLoadedParam
import java.lang.Boolean
import kotlin.Array
import kotlin.Exception
import kotlin.String
import kotlin.Throwable
import kotlin.arrayOf

/**
 * ZUI package installer hook module.
 * Function: bypasses ZUI system installation restrictions and modifies the
 * package installer UI style.
 * Target: com.android.packageinstaller (ZUI system package installer).
 */
@SuppressLint("PrivateApi")
class PackageInstallerStyleHook : AppHookModule() {
    override fun getModuleName(): String = PreferenceKeys.PACKAGE_INSTALLER_STYLE_HOOK.name

    override fun getTargetPackages(): Array<String> = arrayOf(ScopeKeys.PACKAGE_INSTALLER.packageName)

    override fun handleLoadPackage(param: PackageLoadedParam) {
        val classLoader = param.defaultClassLoader
        hookZuiPackageInstaller(classLoader)
        doNotShowWarnTextHook(classLoader)
    }

    private fun hookZuiPackageInstaller(classLoader: ClassLoader) {
        try {
            // 1. Hook Utils.isCTSandGTS to bypass installation restrictions
            hookInstallationRestrictions(classLoader)

            // 2. Hook Activity styles to modify the UI
            hookActivityStyles(classLoader)

            logger.info("ZUI Package Installer Hook loaded successfully")
        } catch (t: Throwable) {
            logger.error("Failed to load ZUI Package Installer Hook", t)
        }
    }

    private fun hookInstallationRestrictions(classLoader: ClassLoader) {
        try {
            val utilsClass = classLoader.loadClass(
                "com.android.packageinstaller.extra.Utils"
            )

            // Hook the overloads of isCTSandGTS
            val isCTSandGTS1 = utilsClass.getDeclaredMethod("isCTSandGTS", String::class.java)
            hookWithId(
                isCTSandGTS1,
                "is_ct_sand_gts1"
            ) { Boolean.TRUE }

            val isCTSandGTS2 =
                utilsClass.getDeclaredMethod("isCTSandGTS", String::class.java, Intent::class.java)
            hookWithId(
                isCTSandGTS2,
                "is_ct_sand_gts2"
            ) { Boolean.TRUE }

            logger.info("Successfully hooked installation restriction checks")
        } catch (t: Throwable) {
            logger.error("Failed to hook installation restriction checks", t)
        }
    }

    private fun hookActivityStyles(classLoader: ClassLoader) {
        try {
            // Get the resource ID of Theme_AlertDialogActivity
            val styleClass = classLoader.loadClass(
                $$"com.android.packageinstaller.R$style"
            )
            val themeField = styleClass.getDeclaredField("Theme_AlertDialogActivity")
            themeField.isAccessible = true
            val themeAlertDialogActivity = themeField.getInt(null)

            // Hook Activity.onCreate to modify the theme and window attributes
            val onCreate = Activity::class.java.getDeclaredMethod("onCreate", Bundle::class.java)
            hookWithId(onCreate, "on_create") { chain ->
                val activity = chain.thisObject as Activity
                // Check whether this is the target package installer Activity
                if (activity.packageName == ScopeKeys.PACKAGE_INSTALLER.packageName) {
                    try {
                        // Apply the dialog theme
                        activity.setTheme(themeAlertDialogActivity)

                        // Enable a transparent background
                        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
                            activity.setTranslucent(true)
                        }

                        // Request no title bar
                        activity.requestWindowFeature(1) // 1 = Window.FEATURE_NO_TITLE

                        // Disable window animations
                        activity.window.setWindowAnimations(0)

                        logger.debug("Successfully modified package installer Activity style")
                    } catch (t: Throwable) {
                        logger.error("Error while modifying Activity style", t)
                    }
                }
                chain.proceed()
            }

            logger.info("Successfully hooked Activity style modification")
        } catch (t: Throwable) {
            logger.error("Failed to hook Activity style modification", t)
        }
    }

    private fun doNotShowWarnTextHook(classLoader: ClassLoader) {
        try {
            val activityClass = classLoader.loadClass(
                "com.android.packageinstaller.PackageInstallerActivity"
            )
            val startInstallConfirm = activityClass.getDeclaredMethod("startInstallConfirm")
            hookWithId(
                startInstallConfirm,
                "start_install_confirm"
            ) { chain ->
                val result = chain.proceed()
                try {
                    val resourcesClass =
                        classLoader.loadClass($$"com.android.packageinstaller.R$id")
                    val warnTextViewIdField =
                        resourcesClass.getDeclaredField("install_confirm_question_warning")
                    warnTextViewIdField.isAccessible = true
                    val warnTextViewId = warnTextViewIdField.getInt(null)

                    val mDialogField =
                        chain.thisObject.javaClass.getDeclaredField("mDialog")
                    mDialogField.isAccessible = true
                    val dialog = mDialogField.get(chain.thisObject) as AlertDialog?

                    val tv = dialog!!.findViewById<TextView>(warnTextViewId)
                    tv.visibility = TextView.GONE
                    logger.debug("Successfully set install warn visibility to GONE")
                } catch (e: Exception) {
                    logger.error("Exception happened when trying to set warn text to GONE!", e)
                }
                result
            }
        } catch (t: Throwable) {
            logger.error("Failed to hook doNotShowWarnTextHook", t)
        }
    }
}
